File: three_sum/three_sum.py
def three_sum(nums):
    nums.sort()
    n = len(nums)
    output = []
    
    for i in range(0, n-2):
        if i > 0 and nums[i] == nums[i -1]:
            continue
        
        target = -nums[i]
        L = i+1;
        R = n-1;
        
        while L < R:
            sum = nums[L] + nums[R]
            if sum == target:
                output.append([nums[i], nums[L], nums[R]])
                L += 1
                R -= 1
                while L < R and nums[L] == nums[L-1]:
                    L += 1
            elif sum < target:
                L += 1
            else:
                R -= 1
    return output
                
                
nums =  [-1, 0, 1, 2, -1, -4]
print("The Three sum of the array is: ",three_sum(nums))

File: three_sum/three_sum.ts
export {};

function three_sum(nums: number[]): number[][] {
  nums.sort((a, b) => a - b);

  const n = nums.length;
  let output: number[][] = [];

  for (let i = 0; i < n-2; i++) {
    if (i > 0 && nums[i] === nums[i - 1]) continue;
    let target = -nums[i];
    let L = i + 1;
    let R = n - 1;

    while (L < R) {
      let sum = nums[L] + nums[R];
      if (sum === target) {
        output.push([nums[i], nums[L], nums[R]]);
        L++;
        R--;
        while (L < R && nums[L] === nums[L - 1]) L++;
      } else if (sum < target) {
        L++;
      } else {
        R--;
      }
    }
  }

  return output;
}

const nums = [-1, 0, 1, 2, -1, -4];
console.log("The Three sum of the array is: ", three_sum(nums));
